sdk/src/stake_weighted_timestamp.rs

```diff
@@ -36,6 +67,48 @@ pub fn calculate_stake_weighted_timestamp(
     }
 }
 
+fn calculate_bounded_stake_weighted_timestamp(
+    unique_timestamps: &HashMap<Pubkey, (Slot, UnixTimestamp)>,
+    stakes: &HashMap<Pubkey, (u64, Account)>,
+    slot: Slot,
+    slot_duration: Duration,
+) -> Option<UnixTimestamp> {
+    let mut stake_per_timestamp: BTreeMap<UnixTimestamp, u128> = BTreeMap::new();
+    let mut total_stake = 0;
+    for (vote_pubkey, (timestamp_slot, timestamp)) in unique_timestamps.iter() {
+        let offset = slot.saturating_sub(*timestamp_slot) as u32 * slot_duration;
+        let estimate = timestamp + offset.as_secs() as i64;
+        let stake = stakes
+            .get(&vote_pubkey)
+            .map(|(stake, _account)| stake)
+            .unwrap_or(&0);
+        stake_per_timestamp
+            .entry(estimate)
+            .and_modify(|stake_sum| *stake_sum += *stake as u128)
+            .or_insert(*stake as u128);
+        total_stake += *stake as u128;
+    }
+    if total_stake == 0 {
+        return None;
+    }
+    let mut stake_accumulator = 0;
+    let mut stake_weighted_timestamps_sum = 0;
+    let mut included_stake = 0;
+    // Only include timestamp estimates from the middle 1/3 stake
+    for (timestamp, stake) in stake_per_timestamp.iter() {
+        stake_accumulator += *stake;
+        if stake_accumulator <= (total_stake / 3) as u128 {
+            continue;
+        }
+        stake_weighted_timestamps_sum += *timestamp as u128 * *stake as u128;
```

Comment:
Even in the middle 1/3 of the stake, I think we should throw away outliers.  I can still sneak in a i64::MAX, perhaps by accident, and cause havoc.  Require timestamps be within an hour of the median?  (even an hour is absurdly generious given all nodes should be synced with NTP time)

Reply:
Agreed. @aeyakovenko proposed tracking the timestamp at the beginning of the epoch and bounding based on the poh estimate from the beginning of the epoch. This seems a little safer to me than a set time-delta from median.